Workflows workshop (#1409)
oktadeveloper/okta-blog: _source/_posts/2023-07-31-workflows-workshop.md

---
layout: blog_post

> title: "Enterprise Ready Workshop: Workflows"

Reviewer: ```suggestion
title: "Enterprise Ready Workshop: Automate with with no-code Okta Workflows"
```

author: max-katz
by: advocate
communities: [no-code,javascript]
description: "Add automated reporting with no-code Okta Workflows. Enhance customer satisfaction by enabling custom automation."
tags: [enterprise-ready-workshops, workflows, no-code, low-code]
tweets:
- ""
image: blog/workflows-workshop/social.jpg
type: awareness
---

> ## Introduction

Reviewer: ```suggestion
```

Reviewer: No need for this

Reviewer: There is no need for this header. It doesn't add any value to the post or SEO. It should be renamed to something that fulfills those two goals

> You built the following Todo application: 

Reviewer: ```suggestion
This tutorial is part of the Enterprise-ready workshop series. In this workshop, you'll enhance the base Todo application by creating an automated report using Okta's no-code Workflows. 

You built the following Todo application: 
```

> {% img blog/workflows-workshop/Workflows_todoapp.jpg alt:"Todo application" %}{: .center-image }

Reviewer: ```suggestion
{% img blog/workflows-workshop/Workflows_todoapp.jpg alt:"The Todo application you built showing tasks to try workflows, watch a workflows video, build your first animation, and read the workflows docs" width:"800" %}{: .center-image }
```

> You need to add one additional enterprise capability. You need to add a todo report that summarizes all the todo items and is automatically emailed once a week. 

Reviewer: ```suggestion
You need to add one additional enterprise capability. You need to add a todo report that summarizes all the todo items for your organization and automatically emails the report once a week. 
```

The email report looks like this: 

> {% img blog/workflows-workshop/Workflows_email_report.jpg alt:"Todo email report" %}{: .center-image }

Reviewer: ```suggestion
{% img blog/workflows-workshop/Workflows_email_report.jpg alt:"Todo email report showing each todo item from the Todo application with a boolean indicating completed state" width:"800" %}{: .center-image }
```

> 

Reviewer: ```suggestion

|Posts in the enterprise-ready workshop series|
| --- |
| 1. [How to get Going with the Enterprise-Ready Identity for SaaS Apps Workshops](/blog/2023/07/27/enterprise-ready-getting-started) |
| 2. [Enterprise-Ready Workshop: Authenticate with OpenID Connect](/blog/2023/07/28/oidc_workshop) |
| 3. [Enterprise-Ready Workshop: Manage Users with SCIM](/blog/2023/07/28/scim-workshop) |
| 4. [Enterprise-Ready Workshop: Terraform](/blog/2023/07/028/terraform-workshop) |
| 5. **Enter-Ready Workshop: Automate with no-code Okta Workflows** |
```

> ## What you need to complete this workshop

Reviewer: ```suggestion
**What you need to complete the workshop**
```

You need access to the following tools to complete this workshop. 

> - [npm](https://www.npmjs.com/)

Reviewer: ```suggestion
```

I think we should remove in favor of pointing people to getting started guide

- The completed todo application ([download](https://github.com/oktadev/okta-enterprise-ready-workshops/tree/oidc-workshop-completed))

> - Access to Okta Worklfows. You will learn how to gain access to Workflows in a later section

Reviewer: ```suggestion
- Access to Okta Worklfows. You will learn how to gain access to Workflows in a later section

You'll also want to ensure you're up and running with the base Todo application by following [How to Get Going with the Enterprise-Ready Identity for SaaS Apps Workshops](/blog/2023/07/27/enterprise-ready-getting-started).
```

{% img blog/workflows-workshop/Workflows_Get_card.jpg alt:"The API Connector - Get card" %}{: .center-image }

You can test the **Get** card. 

1. Click the ▶️ button at the bottom of the card
2. Press Test to test the card and the API service
3. Expand the Body field to see the tasks from the application

{% img blog/workflows-workshop/Workflows_Get_card_testing.jpg alt:"Testing the Get card" %}{: .center-image }

You can also test the flow. Click the **Close** button to go back to flow editing. Click the **Test** button to test the flow. 

The **Flow History** page shows data passed from card to card for each card flow. 

{% img blog/workflows-workshop/Workflows_Get_card_testing_history.jpg alt:"Testing the flow" %}{: .center-image }

In the next section, you will create the text message. 

### Creating the message

In this step, you will create a text message to send via email. Later, you will modify the message to include a list of all the todos. 

1. In the flow editor, click **Add function > Most popular (category) > Compose**
2. In the **Compose** card, enter the following text:

```
Todos Items
```

Your card should look like this: 

> https://katzmax.files.wordpress.com/2023/07/workflows_compose_card.png

Author: ```suggestion
```

{% img blog/workflows-workshop/Workflows_Get_card.jpg alt:"The API Connector - Get card" %}{: .center-image }

You can test the **Get** card. 

1. Click the ▶️ button at the bottom of the card
2. Press Test to test the card and the API service
3. Expand the Body field to see the tasks from the application

{% img blog/workflows-workshop/Workflows_Get_card_testing.jpg alt:"Testing the Get card" %}{: .center-image }

You can also test the flow. Click the **Close** button to go back to flow editing. Click the **Test** button to test the flow. 

The **Flow History** page shows data passed from card to card for each card flow. 

{% img blog/workflows-workshop/Workflows_Get_card_testing_history.jpg alt:"Testing the flow" %}{: .center-image }

In the next section, you will create the text message. 

### Creating the message

In this step, you will create a text message to send via email. Later, you will modify the message to include a list of all the todos. 

1. In the flow editor, click **Add function > Most popular (category) > Compose**
2. In the **Compose** card, enter the following text:

```
Todos Items
```

Your card should look like this: 

> https://katzmax.files.wordpress.com/2023/07/workflows_compose_card.png

Reviewer: @edunham not sure if this image got missed or if the image exists in okta-blog repo now but didn't get converted, but will need an update here.

Author: aha, my script to transform the images that were in markdown format didn't fix it because it wasn't embedded in the dev.to format. Fixing :)

Since all the items were processed, the **Reduce** card is done and the flow execution moves on to the next card, the **Send Email** card. 

{% img blog/workflows-workshop/Workflows_Todo_flow_final_with_Reduce.jpg alt:"The flow continues after the Reduce card processes all the items" %}{: .center-image }

## Testing the flow

To test the flow: 

1. Check that the application's front end, back end, and tunnel are running
    - If the tunnel restarted, you need to update the tunnel URL in the **Get** card
2. Turn both the **Todo Report** and the **Todo Report_Helper** flows on
3. Click the **Test** button on the **Todo Report** flow

You will receive an email like this:

{% img blog/workflows-workshop/Workflows_email_report.jpg alt:"Todo email report" %}{: .center-image }

## What you learned
In this tutorial, you built and learned the following:

- Built your first Workflows flow
- Scheduled the flow to run periodically
- Called an API service to retrieve todos items from an application
- Created a todo summary and sent the summary via email
- Tested the flow

## Workflows resources

> More resources to help you learn more about Okta Workflows.

Reviewer: TODO - OktaDev CTA

Reviewer 2: CTA added.

---
layout: blog_post
title: "Enterprise Ready Workshop: Automate with with no-code Okta Workflows"
author: max-katz
by: advocate
communities: [no-code,javascript]
description: "Add automated reporting with no-code Okta Workflows. Enhance customer satisfaction by enabling custom automation."

> tags: [enterprise-ready-workshops, react, express, oidc, workflows, automation, no-code, low-code]

Reviewer: I believe the tag changes to enterprise maturity? Same with title??? @edunham to weigh in

Author: the tags are still enterprise-ready-workshops on the other posts, so may as well keep it consistent and then I'll batch the change later on

tweets:
- ""
image: blog/workflows-workshop/social.jpg

> type: awareness

Reviewer: I think this is conversion, not awareness. But i don't know if this field matters anymore

Author: I don't think we're using this for anything any more.

---

> This tutorial is part of the enterprise-ready workshop series. In this workshop, you'll enhance the base Todo application by creating an automated report using Okta's no-code Workflows platform.

Reviewer: Please clarify if the term "enterprise-ready workshop series" is correct here

Author: we're calling them "on-demand workshops" now but I haven't updated Terraform and OIDC to match yet... oops!

Reviewer 2: Updated.

You will learn about the Workflows building blocks in a few minutes. First, let's create your first flow. 

## Getting access to Okta Workflows
There are two ways to access Workflows. 

1. You might be entitled to Workflows if you already use some Okta products
2. Access to Workflows as part of WIC (Workforce Identity Cloud) trial

### Workflows when using other Okta products
If you have Okta SSO or Okta UD you might be entitled to Workflows. Check if you can access Workflows by going to **Okta organization > Admin > Workflows > Workflows console**. If you use these products but don't have access to Workflows, please contact your account manager. 

### Okta Workflows as part of WIC (Workforce Identity Cloud) trial
To sign up for a Workforce Identity Cloud trial:

1. Go to the [Okta Workforce trial page](https://www.okta.com/free-trial/workforce-identity/)
2. Register for a trial to access Workflows

### Opening the Okta Workflows console

To access Workflows:

1. Sign-in into your Okta admin dashboard

>     * If you are using a trial account, the admin dashboard link will be in the following format: `https://trial-xxxxxxxx-admin.okta.com/admin/dashboard`

Reviewer: Please change the format of the url here. For example, is it supposed to be 
`https://trial-<OKTA_DOMAIN>-admin.okta.com/admin/dashboard`?

Reviewer 2: Updated.

---
layout: blog_post
title: "Enterprise Ready Workshop: Automate with with no-code Okta Workflows"
author: max-katz
by: advocate
communities: [no-code,javascript]
description: "Add automated reporting with no-code Okta Workflows. Enhance customer satisfaction by enabling custom automation."

> tags: [enterprise-ready-workshops, workflows, no-code, low-code]

Reviewer: @maxkatz please work with @edunham on these tags. they are changing.

Reviewer 2: @edunham please suggest what tags I need to use in the Workflows workshop.

Author: probably crib off of https://raw.githubusercontent.com/oktadev/okta-blog/main/_source/_posts/2023-07-28-oidc_workshop.md? Although it looks like our currently live posts are using the same workshops tag, so @alisaduncan could you update me on what they're changing to?

Reviewer 2: I set the following tags: 
`[enterprise-ready-workshops, react, express, oidc, workflows, automation, no-code, low-code]`

tweets:
- ""
image: blog/workflows-workshop/social.jpg
type: awareness
---

> {% include toc.md %}

Reviewer: This should move to below the first paragraph (or in the intro)

Reviewer: There's also A LOT of headers, so the TOC looks super overwhelming. We need to evaluate when headers are the correct choice, and when to use other forms of separation (such as bold titles for section) or ordered lists, etc.

Reviewer 2: I moved the TOC to the end of the Introduction.

This tutorial is part of the enterprise-ready workshop series. In this workshop, you'll enhance the base Todo application by creating an automated report using Okta's no-code Workflows platform.

## Introduction

You built a SaaS (Software as a Service) Todo application: 

{% img blog/workflows-workshop/Workflows_todoapp.jpg alt:"The Todo application showing five tasks" width:"800" %}{: .center-image }

The application allows entering todo items and marking them completed when done. 

But it's missing an important enterprise capability. 

You need to add a report that summarizes all the Todo items for your organization and automatically emails the report once a week. 

The email report looks like this: 

{% img blog/workflows-workshop/Workflows_email_report.jpg alt:"Todo email report showing each todo item from the Todo application with a boolean indicating completed state" width:"800" %}{: .center-image }

> |Posts in the enterprise-ready workshop series|

Reviewer: The titles changes in this series, so these links need to be updated

Reviewer 2: I removed the Introduction header and updated the titles.

### Sequence actions with logic and timing
Workflows can create non-activated accounts in all apps one week before a new employee's start date and then activate them on their first day.

If an employee leaves your company, Workflows can deactivate the user account, remove their access to all apps except payroll, and then delete the account after a year.

### Send notifications for lifecycle events
For a lifecycle event such as an app assignment or user suspension, Workflows can notify your IT team through email or Slack.

### Log and share lifecycle events
Workflows can query Okta APIs and System Log events, run logic, and even compile data into a CSV file. Then, Workflows can email that file to your teams.

This is a short list of what is possible.

### Workflows flows
In Workflows, you will create flows. A flow is a sequence of steps to complete a goal. It's similar to a script where every code line is a step to complete a goal. In Workflows, you will build the flow visually without writing any code. 

You will learn about the Workflows building blocks in a few minutes. First, let's create your first flow. 

> ## Getting access to Workflows

Reviewer: Should this be Okta Workflows? We need to use the headers as opportunities to reinforce keywords

Reviewer 2: Updated here and other headings.

To sign up for a Workforce Identity Cloud trial:

1. Go to the [Okta Workforce trial page](https://www.okta.com/free-trial/workforce-identity/)
2. Register for a trial to access Workflows

### Opening the Workflows console

To access Workflows:

1. Sign-in into your Okta admin dashboard
    * If you are using a trial account, the admin dashboard link will be in the following format: `https://trial-xxxxxxxx-admin.okta.com/admin/dashboard`
2. On the right-hand side menu, click **Workflows > Workflows console** to open Workflows

You will see the Workflows home page:

{% img blog/workflows-workshop/Workflows_console_main.jpg alt:"Workflows home page" %}{: .center-image }

## Steps to build automated email report using Okta Workflows
In this workshop, you will learn how to build an automated email report in Okta Workflows. 

You will learn and complete the following steps: 

- Building an Okta Workflows flow and automation
- Scheduling the flow to run periodically
- Enhancing an application with a new API service
- Calling the API service to retrieve todos items from an application
- Creating a todo summary and sending the summary via email
- Testing the flow

> 

> This workshop uses a locally deployed application and its API.

Reviewer: Is this on the wrong line? (inadvertent new line?)

Reviewer 2: Deleted.

There are two types of cards:

- App action cards are steps that control other applications or web services. For example, Gmail, Slack, and Jira
- Function cards are steps to interact with, change, or control the data in a flow. Some examples of functions cards:
    - True/False: Evaluate values based on true or false conditions
    -  Flow Control: Manage and manipulate the structure of your flow
    -  Date & Time: Parse and manipulate times and dates
    -  List: Create and iterate over lists of items
    -  Number: Perform mathematical operations
    -  Text: Build, modify, and parse text

Here is again what you are going to build: 

- Calling an API service to retrieve all the todo items
- Creating a message from all the todo items
- Emailing the message
- Set up the flow to run on schedule

Before building the flow, you need to get the Todo application.  

## Setting up the Todo application
In this section, you will complete the following steps:  

1. Getting the Todo application
2. Installing application dependencies
3. Creating the database
4. Adding an organziation to the application
5. Running the application

### Getting the Todo application

> In this step, you will get the Todo application.

Reviewer: We need to link the getting started. There's tool installation reqs that's covered there and not in here

Reviewer 2: Added.

There are two ways to get the Todo application: 

  _Option 1: Clone the repository_
  1. Clone the [Todo application](https://github.com/oktadev/okta-enterprise-ready-workshops/) that already has OIDC support implemented. 
  2. Run the following command `git checkout oidc-workshop-complete`
  

>   _Option 2: Download the application_

Reviewer: There should be some explanation that reader can download the completed application for this tutorial to use with their workflow or something

Reviewer 2: @alisaduncan do you want me to add more information about option #2 (Download the application)?

Reviewer: Like what Nick has in his post right after the Prereqs here
https://developer.okta.com/blog/2022/07/25/express-prisma-tailwind

That way people know there's code for the completed workflows project if they need to compare, or start directly from that step and only add the workflows, etc.

  1. Go to the [okta-enterprise-ready-workshops](https://github.com/oktadev/okta-enterprise-ready-workshops/tree/oidc-workshop-complete) GitHub page
  2. Open the menu for the green **<> Code** button and select **Download Zip**

### Installing application dependencies

1. Navigate to the folder where you cloned or downloaded the Todo application
    * If you downloaded the zip file, unzip the application file
2. Run the following commands to install application dependencies:

```
npm ci
npm install passport-http-bearer
npm install @types/passport-http-bearer -D
```

### Creating the database

Run the following command to generate a database and seed it with two users.  The user's name and password will write to the console.

```
npm run init-db
```

> The users are 

Reviewer: IMO I think instead of writing out the users and passwords here, we should tell people to look in the console for output. In case anything has to change.

Reviewer 2: Updated.

{% img blog/workflows-workshop/Workflows_prisma_db.jpg alt:"Application database" %}{: .center-image }

To create a new organiziation in the application database: 

1. Click on the **Org** model
2. Click **Add record** to add a new record
3. In the **domain** column, enter **matrix.fake**
4. In the **apikey** column, enter **131313** (you might need to scroll to the right to see this column)
5. In the **User** column, click on **0 User** and select both users to add to this organziation (click outside the list to close it)
5. Click **Save 1 change** button to save the new organization and its API key

The organziation record should look like this (some columns are filtered/hidden to show domain, API, and User columns)

{% img blog/workflows-workshop/Workflows_prisma_db_org_record.jpg alt:"Database with new organization" %}{: .center-image }

You are ready to start the application and create the todo items. 

### Running the application

To start the Todo application, run the following command from the application folder: 

```
npm start
```

The create serveral todo items:

1. In a browser, go to http://localhost:3000

> 2. Sign in to the application. Uuse one of the following users:

Reviewer: ```suggestion
2. Sign in to the application. Use one of the following users:
```

Reviewer 2: Corrected.

>     - User: _trinity@whiterabbit.fake_, password: _Zion_

Reviewer: Same comment here about adding the users and passwords directly here vs asking people to look at console.

Reviewer 2: Updated.

    - User: _bob@tables.fake_, password: _correct horse battery staple_
3. Enter several todo items

{% img blog/workflows-workshop/Workflows_todoapp.jpg alt:"Todo application with several items" %}{: .center-image }

You can also use Prisma Studio to view the todo items in the database. 

## Enhancing the application with a new API

In this step, you will add an API service to get all the items belonging to an organization. You will also add authentiation to the API service. 

The automation you will create later in this workshop will use this API service to retrieve the todo items and generate a report. 

1. In the application folder, open the file `apps/api/src/main.ts` in your favorite editor or IDE

### Importing the library for API authentication

> 1. Go to line 7, add the following code:

Reviewer: It's dangerous to have LOC since things can change in the base app (although they shouldn't). You can give approx LOC or say "imports section at the top of the file" etc.

Reviewer 2: Corrected.

``` ts
import passportBearer from 'passport-http-bearer';
```

This module allows authenticating HTTP requests using bearer tokens. 

### Setting up bearer token authentication strategy

> 1. Go to line 59, add the following code: 

Reviewer: Same comment here about specifying the LOC. Is this still correct with the recent updates?

Reviewer 2: I changed this step to add the code at the end of `main.ts` file. Is that okay?

``` ts
// Workshop code start
const BearerStrategy = passportBearer.Strategy;

passport.use(new BearerStrategy(
  async (apikey, done) => {
    const org = await prisma.org.findFirst({
      where: {
        apikey: apikey
      }
    });

    return done(null, org);
  }
));
```

> This code sets up an authentication strategy using an HTTP bearer token. 

Reviewer: Why are we doing this? What is this for? 

And while we're using an API key in this case, why are we not using OAuth client credentials? (there needs to be explanation)

Reviewer 2: @alisaduncan, can you help with the why? API keys are **_simpler_**? But _simpler_ is only sometimes a good reason.

Reviewer: We're using the keys to look up the associated orgs. The way this is set up, it supports scaling across customers.

Reviewer 2: Added.

### Adding the API endpoint for retrieving todo items

1. Right after the code above, add the following code:

``` ts
app.get('/api/org/todos',
  passport.authenticate('bearer'),
  async (req, res) => {
    const todos = await prisma.todo.findMany({
      where: {
        orgId: req.user['id']
      }
    });
    res.json({todos});
});
// Workshop code end
```

This code retrieves all the todo items belonging to an organization. 

Save the changes.

> If you look in the terminal where you started the application, you will see a message that the application detected changes and restarting. 

There is one last step before you start building the automation. In the next section, you will launch an API tunnel. 

> ## Launching the API tunnel

Reviewer: ```suggestion
## Launching the API  in a local tunnel
```

Reviewer 2: Updated.
